src/ser.rs

```diff
@@ -391,10 +394,25 @@ impl<'s> ser::Serializer for &'s Serializer {
 
     fn serialize_newtype_struct<T: ?Sized + Serialize>(
         self,
-        _name: &'static str,
+        name: &'static str,
         value: &T,
     ) -> Result {
-        value.serialize(self)
+        if name == PRESERVED_VALUE_MAGIC {
```

Comment:
Right now, this will perform an actual string comparison for each `name`. Since we know all those strings are compile-time, and duplicate compile-time strings get coalesced even in debug mode, I think we can reduce perf impact here by doing `std::ptr::eq(name, PRESERVED_VALUE_MAGIC)` instead.

Reply:
Unfortunately this doesn't work: it succeeds on release, but fails in debug. (I also tried changing the `const PRESERVED_VALUE_MAGIC` to `static PRESERVED_VALUE_MAGIC` and it still failed... which I find surprising...)

Reply:
Weird, I played in Compiler Explorer before suggesting this and I saw it coalescing strings even in debug mode.

Maybe it gets split up into different object files during monomorphisation of `Serialize`/`Deserialize` though 🤔

It's rather unfortunate to have to compare actual strings on every newtype...

Reply:
I see Rust codegen does check the length first, before comparing bytes, so maybe this is not too bad.